### PR TITLE
#580: Improve error logging of uncaught errors

### PR DESCRIPTION
**Pull request by fregante**
## Related

- https://github.com/pixiebrix/pixiebrix-extension/issues/580
- https://github.com/pixiebrix/pixiebrix-extension/pull/1558

### Don't log errors as awkward "PromiseRejectionEvent"

Before:

Screen Shot 4

After:

Screen Shot 3

### Automatically `reportUncaughtErrors` in all contexts

I think all contexts should report these errors

### Let the global handler report errors instead of using `.catch(reportError)` in some cases

If the promise is already `void` then this extra `.catch` isn't changing the behavior, so it can just be handled by the uncaught error handler.

### CODE REVIEW

File: src/action.tsx [hunk 1]
```diff
@@ -17,9 +17,6 @@
 
 import "@/extensionContext";
 
-// Init rollbar early so we get error reporting on the other initialization
-import "@/telemetry/rollbar";
```
Comment (fregante) [inline src/action.tsx:21]: This is part of `reportUncaughtError`, which is now part of `@/extensionContext` (I think it makes sense to have a single global file with this kind of initialization)

File: src/background/browserAction.ts [hunk 2]
```diff
@@ -52,7 +52,6 @@ async function handleBrowserAction(tab: Tabs.Tab): Promise<void> {
     updateDevTools({ page: `/devtoolsPanel.html?tabId=${tab.id}` });
   } catch (error) {
     await showErrorInOptions("ERR_BROWSER_ACTION_TOGGLE", tab.index);
-    console.error(error);
     reportError(error);
```
Comment (fregante) [inline src/background/browserAction.ts:55]: `reportError` already logs to console, even in the background

File: src/telemetry/reportUncaughtErrors.ts [hunk 3]
```diff
@@ -32,23 +32,15 @@ function defaultErrorHandler(
   }
 
   reportError(errorEvent);
-  errorEvent.preventDefault();
-}
-
-const seen = new WeakSet<ErrorEvent | PromiseRejectionEvent>();
-function avoidLoops(errorEvent: ErrorEvent | PromiseRejectionEvent): void {
```
Comment (fregante) [inline src/telemetry/reportUncaughtErrors.ts:39]: I think this never happened in practice. For this to happen we'd have to `throw` the same error **event** inside `defaultErrorHandler`, which we're not doing.

File: src/telemetry/rollbar.ts [hunk 4]
```diff
@@ -37,7 +40,9 @@ type Payload = {
  *  @see https://docs.rollbar.com/docs/javascript
  *  @see https://docs.rollbar.com/docs/rollbarjs-configuration-reference
  */
-export const rollbar: Rollbar = (() => {
+export const rollbar = initRollbar();
+
+function initRollbar() {
```
Comment (fregante) [inline src/telemetry/rollbar.ts:45]: This gives a little more clarity to the code, showing that code is being run and it's not just exporting a function.

File: src/telemetry/rollbar.ts [hunk 5]
```diff
@@ -136,3 +141,40 @@ export async function updateAuth({
     });
   }
 }
+
+/**
```
Comment (fregante) [inline src/telemetry/rollbar.ts:145]: I merged `telemetry/logging` into this file. No changes were made unless specified.

File: src/telemetry/rollbar.ts [hunk 6]
```diff
@@ -136,3 +141,40 @@ export async function updateAuth({
...
+    rollbar.error(toLogArgument(errorObject));
+    return;
+  }
+
+  // Events are already natively logged
+  if (
+    !(error instanceof ErrorEvent || error instanceof PromiseRejectionEvent)
+  ) {
+    console.error(error);
+  }
```
Comment (fregante) [inline src/telemetry/rollbar.ts:177]: Before, `reportError` was logging all errors and events — unless it was in the background.

Now I just let the events be logged naturally (no `preventDefault()`) and the rest of them are logged regardless of context.

File: src/telemetry/reportUncaughtErrors.ts [hunk 7]
```diff
@@ -32,23 +32,15 @@ function defaultErrorHandler(
   }
 
   reportError(errorEvent);
-  errorEvent.preventDefault();
```
Comment (fregante) [inline src/telemetry/reportUncaughtErrors.ts:35]: There's no point in avoid logging here if we're calling `console.error` later. This just lets the browser log them.

**Review by twschiller — APPROVED** [review]

**Pull request merged**